schism/compiler.ss

```diff
@@ -392,6 +394,34 @@
         (trace-and-error name 'lookup "unbound identifier"))
       ((cdr pair))))
 
+  ;; parse
+  ;; def == (define (name args ...) body)
+  ;;      | (define name body)
+  ;; or return false
+  (define (definition-name def)
+    (and (pair? def)
+         (eq? (car def) 'define)
+         (pair? (cdr def))
+         (if (pair? (cadr def))
+             (caadr def)
+             (cadr def))))
+  (define (function-definition? def)
+    (and (pair? def)
+         (eq? (car def) 'define)
+         (pair? (cdr def))
+         (or (pair? (cadr def))
+             (and (pair? (cddr def))
+                  (pair? (caddr def))
+                  (eq? 'lambda (caaddr def))))))
+  (define (function-formals def)
+    (if (pair? (cadr def))
+        (cdadr def)
+        (car (cdaddr def))))
```

Comment:
For each of these functions, in the `lambda` cause, could you add a check to make sure there is a `'lambda` in the right place? That will help prevent accidental matches as we expand support for more complex expressions.

Reply:
Makes sense! hopefully the added checks look reasonable?

Reply:
Looks good!